### PR TITLE
"getValue() method is redundant, use fromMetaWalEntry instead"

### PR DESCRIPTION
**Pull request by rsingh433**
getValue() method is redundant, the same value can be obtained using the fromMetaWalEntry method.

### CODE REVIEW

File: core/src/test/java/org/apache/accumulo/core/metadata/schema/TabletMetadataTest.java
Comment (keith-turner) [inline core/src/test/java/org/apache/accumulo/core/metadata/schema/TabletMetadataTest.java:111]: Seems like this code could stay the same, is this test failing?

```suggestion
        .put(le2.getValue());
```

File: test/src/main/java/org/apache/accumulo/test/MissingWalHeaderCompletesRecoveryIT.java
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/MissingWalHeaderCompletesRecoveryIT.java:162]: I think this code can stay the same also, let me know if the test is failing and I can take a look at it.


```suggestion
        assertEquals(0, Iterables.size(s));
```

File: test/src/main/java/org/apache/accumulo/test/MissingWalHeaderCompletesRecoveryIT.java
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/MissingWalHeaderCompletesRecoveryIT.java:221]:
```suggestion
        assertEquals(0, Iterables.size(s));
```

File: test/src/main/java/org/apache/accumulo/test/manager/MergeStateIT.java
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/manager/MergeStateIT.java:205]:
```suggestion
     LogEntry logEntry = new LogEntry("localhost:1234/" + UUID.randomUUID());
```

**Review by keith-turner — COMMENTED** [review]
For the test changes, thinking those should not change.  If those test are failing let me know and I would be happy to take a look and see what is happening.

**Review by ctubbsii — COMMENTED** [review]
At a glance, these changes look good to me, but I'm not sure I understand why the test is failing. Or rather, I'm not sure how the test was passing before.

One thing I saw was that the test can be simplified by using a [new SimpleImmutableEntry(key, entry.getValue());](https://docs.oracle.com/javase/8/docs/api/index.html?java/util/AbstractMap.SimpleImmutableEntry.html) rather than creating an anonymous subclass of Entry on lines 60-75.

Another issue is that the LogEntry.validateFilePath method (where this is failing) seems to have an out-of-date concept of the format of the file path. I don't think it's correct.

Given the out of date format that is being validated in validateFilePath method, I see why it's failing (because the test's file name is "foo" instead of a UUID), but I don't understand why it wasn't already failing before.

This will need to be tracked down and understood, I think, in order to make progress here.

**Comment by ctubbsii** [comment]
@keith-turner helped me understand the test failures (or rather, why the current test is passing in the main branch when it probably shouldn't), and I opened up #4035 to help with that. Once that is merged, I'll help resolve the conflicts with this PR, so we can merge this in after that.

**Comment by ctubbsii** [comment]
These changes look good, as-is, and can be merged now to drop storing the "filePath" in the Value and reading it from the column qualifier instead. However, there's some follow-on actions that you can add to this PR, or can create a separate PR for, at your discretion. Here's some things I saw while looking at this code, where there's room for improvement:

1. Pull in the mutation modification work into the LogEntry class. So, instead of calling code calling in to LogEntry to get the column qualifier, value, etc. to update a mutation, you just pass a mutation into the LogEntry class like: `logEntry.addToMutation(mutation);`, and the mutation updating is done internal to that method. That way, the column qualifier, value, etc. doesn't need to be used outside the class (except maybe for unit testing) and they can be made private, or even simplified or removed.
2. Get rid of switchFile method, since that's outdated, and simplify how VolumeUtil is using it
3. Rename "filePath", "filename", and "getFilePath()", etc. to something that reflects what is actually stored there. From what I can tell, what's stored there is a "-/host:port/UUID" (with "-/" ignored). So, maybe a better name is "logReference", since the host and UUID together could be called that. Or something like "serverAndUUID", but that's not really very friendly. If there's a better name, I'm open to that as well.
4. Make the validate method more rigorous and expect exactly 3 items (`"-/host:port/UUID" -> {"-", "host:port", "UUID"}`. Right now validates either 2 or 3... and it's not clear why. Probably so it could validate the old value, which did not contain the "-/" portion, as well as the column qualifier, which did. But it doesn't need to validate the 2-item version, since that shouldn't exist
5. Add upgrade code to drop the "-/" portion entirely, and further simplify all the code that tries to handle that, so it's only "host:port/UUID" instead in the column qualifier.

Some of these are certainly more complicated than others, and would require a separate PR. But some of these are small enough to include with this.

**Comment by rsingh433** [comment]
I removed the switchFile method and updated the switchVolumes method to directly create a new LogEntry instance with the switched file path. Now, the LogEntry class itself is responsible for creating instances based on the file path. Also, I added the functionality to add LogEntry information to a mutation by calling logEntry.addToMutation(mutation).

File: core/src/main/java/org/apache/accumulo/core/tabletserver/log/LogEntry.java
Comment (rsingh433) [reply core/src/main/java/org/apache/accumulo/core/tabletserver/log/LogEntry.java:117]: Thank you. Working on that next.

File: core/src/main/java/org/apache/accumulo/core/tabletserver/log/LogEntry.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/tabletserver/log/LogEntry.java:85]: Could  update the following to use this new method

https://github.com/apache/accumulo/blob/9701ccface1f94b8d4827323b540ddda9d0cd6c6/server/base/src/main/java/org/apache/accumulo/server/metadata/TabletMutatorBase.java#L178

File: core/src/main/java/org/apache/accumulo/core/tabletserver/log/LogEntry.java
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/tabletserver/log/LogEntry.java]: @keith-turner `parts[0]` could be something other than `-`, though. The use of `-` was a backwards-compatible choice to substitute a server name that used to be stored in this field. While it's unlikely that it will contain anything else, it is technically possible if somebody upgrades through multiple versions, since there is no upgrade code to get rid of those. Checking for the length being 2 makes sense, but it doesn't really matter, because it will already fail if it's less than 2, and if it's more than 2 (which is expected, actually), they will be combined into the second one, because we specified "at most 2" as a parameter to the split method. So, all you're doing is adding an extra check to fail with one kind of exception instead of another, and making it less robust on upgrade.

Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/tabletserver/log/LogEntry.java]: #4057 should cover improvements to the validation

**Review by ctubbsii — APPROVED** [review]

File: core/src/main/java/org/apache/accumulo/core/tabletserver/log/LogEntry.java
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/tabletserver/log/LogEntry.java]: Correction, this isn't covered by #4057, because it was included here. I think maybe there's a small risk of things going wrong if there's an old log entry from before 2.1.0 when the `-` started getting used in place of the field that was previously in that position. I created #4061 as follow up.

**Pull request merged**
